taroscript/send.go

```diff
@@ -446,202 +347,301 @@ func CompleteAssetSpend(internalKey btcec.PublicKey, delta SpendDelta,
 
 	// If the transfer contains no asset splits, we only need to validate
 	// the new asset with its witness attached.
-	if updatedDelta.SplitCommitment == nil {
+	if !input.IsSplit {
 		if err := verifySpend(nil); err != nil {
-			return nil, err
+			return err
 		}
 
-		updatedDelta.NewAsset = *validatedAsset
-
-		return &updatedDelta, nil
+		return nil
 	}
 
 	// If the transfer includes an asset split, we have to validate each
 	// split asset to ensure that our new Asset is committing to
 	// a valid SplitCommitment.
-	splitAssets := maps.Values(updatedDelta.SplitCommitment.SplitAssets)
-	err = verifySpend(splitAssets)
-	if err != nil {
-		return nil, err
+	splitAssets := make([]*commitment.SplitAsset, 0, len(outputs)-1)
+	for idx := range outputs {
+		if outputs[idx].IsChange || outputs[idx].Interactive {
+			continue
+		}
+
+		splitAssets = append(splitAssets, &commitment.SplitAsset{
+			Asset:       *outputs[idx].Asset,
+			OutputIndex: outputs[idx].AnchorOutputIndex,
+		})
+	}
+	if err := verifySpend(splitAssets); err != nil {
+		return err
 	}
 
 	// Update each split asset to store the root asset with the witness
 	// attached, so the receiver can verify inclusion of the root asset.
-	for key := range updatedDelta.SplitCommitment.SplitAssets {
-		updatedDelta.SplitCommitment.SplitAssets[key].Asset.
-			PrevWitnesses[0].SplitCommitment.
-			RootAsset = *validatedAsset.Copy()
-	}
+	for idx := range outputs {
+		if outputs[idx].IsChange || outputs[idx].Interactive {
```

Comment:
`splitAssets` should also include the split root / change output.

Related, I think `Interactive` and `!IsSplit` are mutually exclusive? At least while we only have one input.

Reply:
Confirmed that the VM expects the split asset representing the root to be present in this list:

https://github.com/lightninglabs/taro/blob/82daf05d4fd50d14e47010365c14a1416e806350/vm/vm.go#L152

And we pass the full list in tests:

https://github.com/lightninglabs/taro/blob/82daf05d4fd50d14e47010365c14a1416e806350/vm/vm_test.go#L266

IIUC the root split asset is the only place storing the merkle proof of inclusion for the root asset, so we should still be verifying it.

Reply:
Hmm, I'm not sure about this. We don't have that split asset anymore here, it's discarded in `PrepareOutputAsset`, as we only need the root asset (which is a copy of the split asset at the root locator, with the tx witness and split commitment root set). It's true that the VM will validate it if supplied. But it doesn't require it, and all the proofs seem to work without it (in the old code we only ever used that split asset in this verify call after signing, it wasn't used anywhere else afterwards).
So because I think it's easier to understand what happens, and because the VM doesn't require it, it's easier to just drop it?

If there was crucial information missing, then the proofs wouldn't validate, right?

Reply:
Yeah, I think the lack of failures is a related issue where, from the VM, we can't know that we verified all leafs of of the split tree, only that there is non-inflation.

AFAICT we don't use the proof-of-inclusion for the root asset later, and running `validateSplit()` with the root locator doesn't add any validation for the root besides that inclusion proof.

If we are dropping the root locator then `validateSplit()` needs a change to enforce non-zero amounts for splits, but that can be a separate PR.

Reply:
@Roasbeef does this sound correct?

Reply:
Including the root there was meant to allow for the root + all splits to be validated in a single go. 

> Hmm, I'm not sure about this. We don't have that split asset anymore here, it's discarded in PrepareOutputAsset, as we only need the root asset (which is a copy of the split asset at the root locator, with the tx witness and split commitment root set).

Sounds like y'all are referring to the same thing here? The root asset here is the same as "change" in this parlance, which is the asset that anchors all the created external splits.

Reply:
We talked about it again offline and I think we are in sync now. I added a commit (7170c4a8) to solidify this updated assumption in the VM. I hope you agree with the change as well, @Roasbeef.

Reply:
From offline discussion:

Modify the vInput format to include an optional field for the merkle proof that the root asset is included in the split commitment tree. Also update encode/decode to include this field, and update `SignVirtualTransaction` to pass this proof to the VM in the existing form.